src/notebookConcatDocument.ts

```diff
@@ -1,356 +1,334 @@
+
+    private dummyUri: Uri;
+
+    private _id = uuid();
+
+    private onDidChangeSubscription: Disposable;
+
+    private cellTracking: { uri: Uri; lineCount: number; length: number }[] = [];
+
+    private onCellsChangedEmitter = new EventEmitter<TextDocumentChangeEvent>();
+
+    private _notebook: NotebookDocument;
+    private _selector: string;
+
+    constructor(
+        notebook: NotebookDocument,
+        notebookApi: IVSCodeNotebook,
+        selector: string,
+        public readonly key: string
+    ) {
+        const dir = path.dirname(notebook.uri.fsPath);
+        this._selector = selector;
+        // Create a safe notebook document so that we can handle both >= 1.56 vscode API and < 1.56
+        // when vscode stable is 1.56 and both Python release and insiders can update to that engine version we
+        // can remove this and just use NotebookDocument directly
+        this._notebook = notebook;
+        // Note: Has to be different than the prefix for old notebook editor (HiddenFileFormat) so
+        // that the caller doesn't remove diagnostics for this document.
+        this.dummyFilePath = path.join(dir, `${NotebookConcatPrefix}${uuid().replace(/-/g, '')}.py`);
+        this.dummyUri = Uri.file(this.dummyFilePath);
+
+        if (notebook.uri.scheme === InteractiveScheme) {
+            this.concatDocument = new InteractiveConcatTextDocument(notebook, selector, notebookApi);
+        } else {
+            this.concatDocument = new EnhancedNotebookConcatTextDocument(notebook, selector, notebookApi);
+        }
+
+        this.onDidChangeSubscription = this.concatDocument.onDidChange(this.onDidChange, this);
+        this.updateCellTracking();
+    }
+
+    notebook: NotebookDocument | undefined = undefined;
+
+    public dispose(): void {
+        this.onDidChangeSubscription.dispose();
+        this.onCellsChangedEmitter.dispose();
+    }
+
+    public get id() {
+        return this._id;
+    }
+
+    public isCellOfDocument(uri: Uri): boolean {
+        return this.concatDocument.contains(uri);
+    }
+
+    // eslint-disable-next-line class-methods-use-this
+    public save(): Thenable<boolean> {
+        // Not used
+        throw new Error('Not implemented');
+    }
+
+    public lineAt(posOrNumber: Position | number): TextLine {
+        return this.concatDocument.lineAt(posOrNumber);
+    }
+
+    public offsetAt(position: Position): number {
+        return this.concatDocument.offsetAt(position);
+    }
+
+    public positionAt(offset: number): Position {
+        return this.concatDocument.positionAt(offset);
+    }
+
+    public getText(range?: Range | undefined): string {
+        const concatText = range ? this.concatDocument.getText(range) : this.concatDocument.getText();
+
+        // Concat document doesn't put the extra newline at the end of the last cell. This prevents
+        // cell additions after an open from being concated correctly
+        if (concatText.endsWith('\n')) {
```

Comment:
This was causing problems based on when the docOpen message was being sent. Pylance would get a textDocument/change message that didn't include the '\n' in between the initial contents and any subsequent cells.